qiskit_experiments/test/iq_pulse_backend.py

```diff
@@ -0,0 +1,195 @@
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""A Pulse simulation backend based on Qiskit-Dynamics"""
+import datetime
+from typing import Union
+
+import numpy as np
+from qiskit import QuantumCircuit
+from qiskit.providers import BackendV2
+from qiskit.providers.models import PulseDefaults
+from qiskit.providers.options import Options
+from qiskit.qobj.utils import MeasLevel, MeasReturnType
+from qiskit.quantum_info.states import Statevector
+from qiskit.result import Result
+from qiskit.transpiler import Target
+
+from qiskit_dynamics import Solver
+from qiskit_dynamics.pulse import InstructionToSignals
+
+from qiskit_experiments.exceptions import QiskitError
+from qiskit_experiments.test.utils import FakeJob
+
+
+class IQPulseBackend(BackendV2):
+    """Pulse Simulator abstract class"""
+
+    def __init__(self, static_hamiltonian, hamiltonian_operators, dt=0.1 * 1e-9, **kwargs):
+        """Hamiltonian and operators is the Qiskit Dynamics object"""
+        super().__init__(
+            None,
+            name="PulseBackendV2",
+            description="A PulseBackend simulator",
+            online_date=datetime.datetime.utcnow(),
+            backend_version="0.0.1",
+        )
+        self._defaults = PulseDefaults.from_dict({})
+        self.converter = None
+
+        self.static_hamiltonian = static_hamiltonian
+        self.hamiltonian_operators = hamiltonian_operators
+        self.solver = Solver(self.static_hamiltonian, self.hamiltonian_operators, **kwargs)
+        self._target = Target(dt=dt, granularity=16)
+        self.gound_state = np.zeros(self.solver.model.dim)
+        self.gound_state[0] = 1
+        self.y_0 = np.eye(self.solver.model.dim)
+
+    @property
+    def target(self):
+        return self._target
+
+    @property
+    def max_circuits(self):
+        return None
+
+    def defaults(self):
+        """return backend defaults"""
+        return self._defaults
+
+    @classmethod
+    def _default_options(cls):
+        return Options(shots=1024)
+
+    @staticmethod
+    def _get_info(inst):
+        p_dict = inst.operation
+        qubit = tuple(int(str(val)[-2]) for val in inst.qubits)
+        params = tuple(float(val) for val in p_dict.params)
+        return qubit, params, p_dict.name
+
+    @staticmethod
+    def _state_vector_to_result(
+        state: Union[Statevector, np.ndarray],
+        shots: int,
+        meas_return: MeasReturnType,
+        meas_level: MeasLevel,
+    ):
+        """Convert the state vector to IQ data or counts."""
+
+        # if meas_level == MeasLevel.CLASSIFIED:
+        # sample from the state vector. There might already be functions in Qiskit to do this.
+        #     counts = ...
+        # if meas_level == MeasLevel.KERNELED:
+        #     iq_data = ... create IQ data.
+        pass
+
+    def solve(self, schedule_blocks, qubits):
```

Comment:
```suggestion

    @lru_cache
    def solve(self, schedule_blocks, qubits):
```
This should avoid simulating the same schedule multiple times. Note that we will need to amend the imports.

Reply:
done